src/dbxincluder/xinclude.py

```diff
@@ -0,0 +1,305 @@
+    validate_xinclude(elem, file)
+
+    # Get base (nearest xml:base or current directory)
+    base_urls = elem.xpath("ancestor-or-self::*[@xml:base][1]/@xml:base")
+
+    base_url = base_urls[0] if len(base_urls) == 1 else base_url
+    if base_url is None:
+        raise DBXIException(elem, "Could not get base URL", file)
+
+    # Load target
+    try:
+        content, url = get_target(elem, base_url, file)
+    except ResourceError as rex:
+        if not handle_xifallback(elem, file, xinclude_stack):
+            raise rex
+
+        # Is this output appropriate?
+        sys.stderr.write(str(rex) + "\n")
+        return
+
+    # Save text after element
+    saved_tail = elem.tail
+    elem.tail = ""
+
+    # Include as text
+    if elem.get("parse", "xml") != "xml":
+        prev = elem.getprevious()
+        if prev is not None:
+            prev.tail += str(content, encoding="utf-8") + saved_tail
+        else:
+            elem.getparent().text += str(content, encoding="utf-8") + saved_tail
+        return
+
+    # Check for infinite recursion
+    if xinclude_stack is None:
+        xinclude_stack = []
+
+    fragid = elem.get("fragid")
+    xinclude_id = "{0!r}>{1!r}".format(url, fragid)
+    if xinclude_id in xinclude_stack:
+        raise DBXIException(elem, "Infinite recursion detected", file)
+
+    # Parse as XML
+    subtree = fromstring(content)
+
+    # Get subdocument
+    if fragid is not None:
+        subtree = subtree.xpath("//*[@xml:id={0!r}]".format(fragid))
+        if len(subtree) == 1:
+            subtree = subtree[0]
+            # Get xml:base of subdocument
+            base_urls = subtree.xpath("ancestor-or-self::*[@xml:base][1]/@xml:base")
+            url = base_urls[0] if len(base_urls) == 1 else url
+        else:
+            raise DBXIException(elem, file=file,
+                                message="Could not find fragid {0!r} in target {1!r}"
+                                .format(fragid, url))
+
+    # Copy certain attributes from xi:include to the target tree
+    copy_attributes(elem, subtree)
+
+    subtree.tail = saved_tail
+    process_tree(subtree, url, url, xinclude_stack + [xinclude_id])
+
+    # Replace XInclude by subtree
+    elem.getparent().replace(elem, subtree)
+
+
+def process_subtree(tree, base_url, file, xinclude_stack):
+    """Like process_tree, but for subtrees."""
+
+    # for elem in tree.getiterator() does not work here, as we modify tree in-place
+    for elem in tree:
+        if not isinstance(elem.tag, str):
+            continue
+
+        if QName(elem) == QN["xi:include"]:
+            handle_xinclude(elem, base_url, file, xinclude_stack)
+            # handle_xinclude calls process_tree itself if required
+        else:
+            process_subtree(elem, base_url, file, xinclude_stack)
+
+
+def flatten_subtree(tree):
+    """Remove all xi:fallback elements in tree by replacing them with their content."""
+
+    i = 0
+    while i < len(tree):
```

Comment:
Solved: ```flatten_subtree``` was called way to often.
I changed it so that flatten_subtree is only called once now, after the actual xi:include processing.

Now your snippet breaks the tests :P